src/django_otp/plugins/otp_email/models.py

```diff
@@ -57,7 +59,27 @@ def generate_challenge(self, extra_context=None):
         :type extra_context: dict
 
         """
+        generate_allowed, data_dict = self.generate_is_allowed()
+        if generate_allowed:
+            message = self._deliver_token(extra_context)
+
+        else:
+            if data_dict['reason'] == 'COOLDOWN_DURATION_PENDING':
+                next_generation_naturaltime = naturaltime(
+                    data_dict['next_generation_at']
+                )
+                message = (
+                    "Token generation cooldown period has not expired yet. Next"
+                    f" generation allowed {next_generation_naturaltime}."
+                )
+            else:
+                message = "Token generation is not allowed at this time"
+
+        return message
+
+    def _deliver_token(self, extra_context):
         self.generate_token(valid_secs=settings.OTP_EMAIL_TOKEN_VALIDITY)
+        self.last_generated_timestamp = timezone.now()
```

Comment:
This never gets saved. As importantly, the tests aren't catching this (see companion note in the tests). 

(Hint: try swapping these two lines. Adding the technically redundant `commit=True` will help to highlight the fact that you're relying on `generate_token` to save the model instance for you.)